src/main/scala/com/fulcrumgenomics/alignment/Matrix.scala

```diff
@@ -0,0 +1,101 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.alignment
+
+import com.fulcrumgenomics.FgBioDef._
+
+import scala.reflect.ClassTag
+
+/** Factory method(s) for Matrices. */
+object Matrix {
+  def apply[A: ClassTag](x: Int, y: Int): Matrix[A] = {
+    require(x >= 0, s"X dimension of matrix ($x) is less than 0.")
+    require(y >= 0, s"Y dimension of matrix ($y) is less than 0.")
+    new LinearMatrix(x, y)
+  }
+}
+
+
+/**
+  * Defines methods applicable to a 2D matrix.  Matrix cells are accessed with a pair of
+  * coordiantes, the first of which is the row offset starting at 0, and the second of
+  * which is the column offset, starting at 0.
+  * */
+trait Matrix[A] {
+  /** Access a cell in the matrix. */
+  def apply(i: Int, j: Int): A
+
+  /** Updates a cell in the matrix. */
+  def update(i: Int, j: Int, value: A): Unit
+
+  /**
+    * Returns the dimensions of the matrix.  E.g. a 2x2 matrix will return (2,2), although the maximum
+    * valid offsets are (1,1).
+    */
+  def dimensions: (Int, Int)
+
+  /** Returns the total number of cells in the matrix. */
+  def size: Int = {
+    val (x,y) = dimensions
+    x * y
+  }
+
+  /** Provides a simplistic text dump of the matrix, mostly for debugging. */
+  override def toString: String = {
+    val (x,y) = dimensions
+    val buffer = new StringBuilder
+    forloop (from=0, until=x) { i =>
+      if (i > 0) buffer.append("\n")
+
+      forloop (from=0, until=y) { j =>
+        if (j > 0) buffer.append("\t")
+        buffer.append(String.valueOf(this(i,j)))
+      }
+    }
+
+    buffer.toString()
+  }
+}
+
+
+/** Implements a matrix using a single linear array. */
+class LinearMatrix[A: ClassTag](val x: Int, val y: Int) extends Matrix[A] {
+  private val data = new Array[A](x * y)
+
+  /** Access a cell in the matrix. */
+  override def apply(i: Int, j: Int): A = {
+    if (i >= x || j >= y) throw new NoSuchElementException(s"Illegal matrix indices: (${i}, ${j})")
+    data(i*y + j)
+  }
+
+  /** Updates a cell in the matrix. */
+  override def update(i: Int, j: Int, value: A): Unit = {
+    if (i >= x || j >= y) throw new NoSuchElementException(s"Illegal matrix indices: (${i}, ${j})")
+    data(i*y + j) = value
+  }
+
+  /** Returns the dimensions of the matrix. */
+  override def dimensions: (Int, Int) = (x, y)
```

Comment:
It's not needed, but it is helpful.  E.g. it will cause a compile error if it does not match the method name in the trait (e.g. because the trait removed that method).